File: src/main/java/com/dongfg/project/api/service/MessageService.java
package com.dongfg.project.api.service;

import com.alibaba.fastjson.JSON;
import com.dongfg.project.api.component.SmsComponent;
import com.dongfg.project.api.component.XingeComponent;
import com.dongfg.project.api.component.webpush.WebPushComponent;
import com.dongfg.project.api.dto.CommonResponse;
import com.dongfg.project.api.graphql.payload.MessagePayload;
import com.dongfg.project.api.graphql.type.Message;
import com.dongfg.project.api.graphql.type.PushSubscription;
import com.dongfg.project.api.repository.PushSubscriptionRepository;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

/**
 * @author dongfg
 * @date 18-1-3
 */
@Service
@Slf4j
public class MessageService {

    @Autowired
    private PushSubscriptionRepository pushSubscriptionRepository;

    @Autowired
    private SmsComponent smsComponent;

    @Autowired
    private WebPushComponent webPushComponent;

    @Autowired
    private XingeComponent xingeComponent;


    public MessagePayload sendMessage(Message input) {
        MessagePayload messagePayload = MessagePayload.builder().success(true).build();
        CommonResponse commonResponse = null;
        switch (input.getType()) {
            case APP:
                commonResponse = xingeComponent.sendMessage(input);
                messagePayload.setSuccess(commonResponse.isSuccess());
                messagePayload.setMsg(commonResponse.getMsg());
                break;
            case SMS:
                commonResponse = smsComponent.sendMessage(input.getReceiver(), input.getContent());
                messagePayload.setSuccess(commonResponse.isSuccess());
                messagePayload.setMsg(commonResponse.getMsg());
                break;
            case WEB:
                pushSubscriptionRepository.findAll().forEach(subscription ->
                        webPushComponent.sendMessage(subscription, JSON.toJSONString(input)));
                break;
            default:
        }

        return messagePayload;
    }

    public MessagePayload savePushSubscription(PushSubscription subscription) {
        pushSubscriptionRepository.save(subscription);
        return MessagePayload.builder().success(true).build();
    }
}

File: src/main/kotlin/com/dongfg/project/api/service/MessageService.kt
package com.dongfg.project.api.service

import com.dongfg.project.api.common.Constants.MessageType.*
import com.dongfg.project.api.component.Sms
import com.dongfg.project.api.model.Message
import com.dongfg.project.api.web.payload.GenericPayload
import mu.KLogging
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class MessageService {
    companion object : KLogging()

    @Autowired
    private lateinit var sms: Sms

    fun sendMessage(message: Message): GenericPayload {
        val payload = GenericPayload(true)
        when (message.type) {
            SMS -> {
                val result = sms.send(message.receiver, message.content)
                payload.success = result.isSucc
                payload.msg = result.msg
            }
            WEB -> TODO()
            APP -> TODO()
            EMAIL -> TODO()
        }

        return payload
    }
}